### PR TITLE
[DataGrid] Add WCAG 2.2 single-click column resize capability

### PR DESCRIPTION
**Pull request by vnbaaij**
**Do not auto-merge. Need to clear out IssueTester before.**

**Update:** the `three-dot/more` icon has been replaced with the chevron down icon. Also, all header icons have an opacity set now to make them a bit 'softer'

image

Allows for resizing columns in a DataGrid by using the column options menu. Two different modes are available: discrete and exact.
- Discrete: resize the column is steps of -10/+10 pixels (same as +/- keyboard resizing)
- Exact: specify the column width in pixels

Both options offer a reset button to go back to the original grid setting (same as shift+r keyboard reset)

As the options menu shows more then just a column filter, the icon to indicate an option menu is available is changed to the three dots. The indicator for an applied filter has been moved to be shown besides the title akin to the sort icon.

### Resize discrete:
datagrid-resize-discrete
### Resize direct:
datagrid-resize-exact

### CODE REVIEW

File: examples/Demo/Shared/Pages/DataGrid/Examples/DataGridTypical.razor [hunk 1]
```diff
@@ -1,21 +1,41 @@
 ﻿@inject DataSource Data
 
+<p>To test set ResizeType on the DataGrid to either DataGridResizeType.Discrete or DataGridResizeType.Exact</p>
+<p>Remove the parameter completely to get the original behavior</p>
 
-<FluentDataGrid Items="@FilteredItems" ResizableColumns=true Pagination="@pagination" GridTemplateColumns="0.2fr 1fr 0.2fr 0.2fr 0.2fr 0.2fr" RowClass="@rowClass" RowStyle="@rowStyle" Style="height: 405px;overflow:auto;">
+<FluentDataGrid Items="@FilteredItems" ResizableColumns=true ResizeType="DataGridResizeType.Exact" Pagination="@pagination" GridTemplateColumns="0.2fr 1fr 0.2fr 0.2fr 0.2fr 0.2fr" RowClass="@rowClass" RowStyle="@rowStyle" Style="height: 405px;overflow:auto;">
```
Comment (dvoituron) [inline examples/Demo/Shared/Pages/DataGrid/Examples/DataGridTypical.razor:6]: The use of `ResizeType="DataGridResizeType.Discrete"` fails when the increment buttons are clicked.

peek_3

File: src/Core/Components/DataGrid/Columns/ColumnBase.razor [hunk 2]
```diff
@@ -17,62 +17,102 @@
...
                     {
-                        <FluentIcon Value="@(new CoreIcons.Regular.Size24.Filter())" />
+                        <FluentButton Appearance="Appearance.Stealth" class="col-options-button" @onclick="@(() => Grid.ShowColumnOptionsAsync(this))" aria-label="Filter this column">
+                            @if (Filtered.GetValueOrDefault())
+                            {
+                                <FluentIcon Value="@(new CoreIcons.Regular.Size24.FilterDismiss())" />
+                            }
+                            else
+                            {
+                                    <FluentIcon Value="@(new CoreIcons.Regular.Size24.Filter())" />
```
Comment (dvoituron) [inline src/Core/Components/DataGrid/Columns/ColumnBase.razor:40]: This line is not correctly aligned.

File: src/Core/Components/DataGrid/Columns/ColumnResizeOptions.razor [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+@typeparam TGridItem
+
+<FluentStack Class="resize-options" VerticalAlignment="VerticalAlignment.Center" HorizontalGap="4">
+    @if (ResizeType == DataGridResizeType.Discrete)
+    {
+        <FluentLabel>@Label</FluentLabel>
+
+        <FluentSpacer />
+        <FluentButton OnClick="HandleShrinkAsync">
+            <FluentIcon Value="@(new CoreIcons.Regular.Size20.Subtract())" />
```
Comment (dvoituron) [inline src/Core/Components/DataGrid/Columns/ColumnResizeOptions.razor:11]: Could you use `FluentButton.IconStart`?

File: src/Core/Components/DataGrid/Columns/ColumnResizeOptions.razor [hunk 4]
```diff
@@ -0,0 +1,45 @@
+﻿@namespace Microsoft.FluentUI.AspNetCore.Components
+@typeparam TGridItem
+
+<FluentStack Class="resize-options" VerticalAlignment="VerticalAlignment.Center" HorizontalGap="4">
+    @if (ResizeType == DataGridResizeType.Discrete)
+    {
+        <FluentLabel>@Label</FluentLabel>
+
+        <FluentSpacer />
+        <FluentButton OnClick="HandleShrinkAsync">
```
Comment (dvoituron) [inline src/Core/Components/DataGrid/Columns/ColumnResizeOptions.razor:10]: Please, for clean code, could you use the `@` sign to inform the user that this is a .NET method? `@HandleShrinkAsync`

File: src/Core/Components/DataGrid/Columns/ColumnResizeOptions.razor.cs [hunk 5]
```diff
@@ -0,0 +1,67 @@
+using Microsoft.AspNetCore.Components;
+using Microsoft.AspNetCore.Components.Web;
+using Microsoft.FluentUI.AspNetCore.Components.DataGrid.Infrastructure;
+
+namespace Microsoft.FluentUI.AspNetCore.Components;
+public partial class ColumnResizeOptions<TGridItem>
+{
+    private string? _width;
+    [CascadingParameter] internal InternalGridContext<TGridItem> InternalGridContext { get; set; } = default!;
```
Comment (dvoituron) [inline src/Core/Components/DataGrid/Columns/ColumnResizeOptions.razor.cs:9]: Can you write the attribute above the method (another line)?

File: src/Core/Components/DataGrid/Columns/ColumnResizeOptions.razor.cs [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+    /// <summary>
+    /// Gets a reference to the enclosing <see cref="FluentDataGrid{TGridItem}" />.
+    /// </summary>
+    public FluentDataGrid<TGridItem> Grid => InternalGridContext.Grid;
```
Comment (dvoituron) [inline src/Core/Components/DataGrid/Columns/ColumnResizeOptions.razor.cs:14]: Can you use `private` ?

File: src/Core/Components/DataGrid/Columns/ColumnResizeOptions.razor.cs [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+
+    [Parameter]
+    public int Column { get; set; }
```
Comment (dvoituron) [inline src/Core/Components/DataGrid/Columns/ColumnResizeOptions.razor.cs:17]: Can you add the documentation ?

File: src/Core/Components/DataGrid/FluentDataGrid.razor.cs [hunk 8]
```diff
@@ -69,6 +69,17 @@ public partial class FluentDataGrid<TGridItem> : FluentComponentBase, IHandleEve
...
+    /// <summary>
+    /// To comply with WCAG 2.2, a one-click option should be offered to change column widths. We providesuch an option through the
+    /// ColumnOptions UI. This parameter allows you to enable or disable this resize UI.
+    /// Defualts to false.
+    /// </summary>
+    [Parameter]
+    public DataGridResizeType? ResizeType { get; set; }
+
+    [Parameter]
+    public string ResizeLabel { get; set; } = "Column width (in pixels)";
```
Comment (dvoituron) [inline src/Core/Components/DataGrid/FluentDataGrid.razor.cs:81]: Can you add the documentation?

**Review by dvoituron — CHANGES_REQUESTED** [review]

File: src/Core/Components/DataGrid/FluentDataGrid.razor.cs [hunk 9]
```diff
@@ -69,6 +69,17 @@ public partial class FluentDataGrid<TGridItem> : FluentComponentBase, IHandleEve
     /// </summary>
     [Parameter] public bool ResizableColumns { get; set; }
 
+    /// <summary>
+    /// To comply with WCAG 2.2, a one-click option should be offered to change column widths. We providesuch an option through the
```
Comment (dvoituron) [inline src/Core/Components/DataGrid/FluentDataGrid.razor.cs:73]: The text must be fixed "providesuch", "Defualts to false" (not a Boolean) :-)

**Review by dvoituron — CHANGES_REQUESTED** [review]

File: src/Core/Components/DataGrid/Columns/ColumnBase.razor [hunk 10]
```diff
@@ -17,62 +17,102 @@
...
-                <FluentButton Appearance="Appearance.Stealth" class="col-options-button" @onclick="@(() => Grid.ShowColumnOptionsAsync(this))" aria-label="Filter this column">
-                    @if (Filtered.GetValueOrDefault())
-                    {
-                        <FluentIcon Value="@(new CoreIcons.Regular.Size24.FilterDismiss())" />
-                    }
-                    else
+                @if (Grid.ResizeType is not null)
+                {
+                    <FluentButton Appearance="Appearance.Stealth" class="col-options-button" @onclick="@(() => Grid.ShowColumnOptionsAsync(this))" aria-label="Filter this column">
+                        <FluentIcon Value="@(new CoreIcons.Regular.Size24.ChevronDown())" Color="Color.Neutral" Style="opacity: 0.5;" />
```
[the remark this answers is not in the document]
Comment (vnbaaij) [reply src/Core/Components/DataGrid/Columns/ColumnBase.razor:25]: 16px:
image

Think 16 is a bit too small. How about 20?

20px: 
image

**Review by dvoituron — APPROVED** [review]

**Pull request merged**